### PR TITLE
Guards for cuda enabled RAJA

### PR DESCRIPTION
**Pull request by artv3**
By adding a few checks, this PR enables users to maintain one build of RAJA and be compatible with strict device code and CUDA code. This was motivated by an application which compiled packages which depended on RAJA with both nvcc and host compilers.

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
@artv3 - looks like this breaks the clang-cuda builds, I think you might need a different macro. See [this](https://llvm.org/docs/CompileCudaWithLLVM.html#detecting-clang-vs-nvcc-from-code).

File: include/RAJA/index/ListSegment.hpp [hunk 1]
```diff
@@ -38,7 +38,7 @@
 #include "RAJA/util/macros.hpp"
 #include "RAJA/util/types.hpp"
 
-#if defined(RAJA_ENABLE_CUDA)
+#if defined(__NVCC__) && defined(RAJA_ENABLE_CUDA)
```
Comment (davidbeckingsale) [inline include/RAJA/index/ListSegment.hpp]: Probably want `__CUDACC__` rather than `__NVCC__`

Comment (artv3) [reply include/RAJA/index/ListSegment.hpp]: Ah even better! Thanks.

Comment (davidbeckingsale) [reply include/RAJA/index/ListSegment.hpp]: Your solution looks good too.

Comment (artv3) [reply include/RAJA/index/ListSegment.hpp]: Not sure why, but the  ``#if defined(__CUDACC__) && defined(RAJA_ENABLE_CUDA)``  combination didn't work. I reverted to the previous commit.

Comment (davidbeckingsale) [reply include/RAJA/index/ListSegment.hpp]: Works for me.

**Pull request merged**